### PR TITLE
Created a Product model and controller

### PR DESCRIPTION
**Pull request by Sara-mersha**

### CODE REVIEW

File: database/migrations/2024_09_18_084403_create_products_table.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            $table->decimal('price', 10, 2);
+            $table->boolean('flexible_pricing')->default(false);
+            $table->string('brand')->nullable();
+            $table->string('model')->nullable();
+            $table->json('image_urls')->nullable(); // Storing images as JSON array
+            $table->string('barcode_upc')->nullable();
+            $table->string('barcode_eac')->nullable();
+            $table->boolean('product_availability')->default(true);
+            $table->json('tags')->nullable();
+            $table->unsignedBigInteger('company_id');
```
Comment (BinyamMamo) [inline database/migrations/2024_09_18_084403_create_products_table.php]: `company_id` should be renamed to `vendor_id` for consistency

File: app/Http/Controllers/ProductController.php [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    // Create a new product
+    public function store(Request $request)
+    {
+        $validatedData = $request->validate([
+            'title' => 'required|string|max:255',
+            'description' => 'nullable|string',
+            'price' => 'required|numeric',
+            'flexible_pricing' => 'required|boolean',
```
Comment (BinyamMamo) [inline app/Http/Controllers/ProductController.php]: `flexible_pricing` should be nullable, because it is `false` by default

**Comment by carnage-mode** [comment]
For now I've removed the update functionality because I wasn't quite sure what we should allow updating, we should investigate that separately

File: database/migrations/2024_09_18_084403_create_products_table.php (hunk 1 above)
Comment (carnage-mode) [reply database/migrations/2024_09_18_084403_create_products_table.php]: done

File: app/Http/Controllers/ProductController.php (hunk 2 above)
Comment (carnage-mode) [reply app/Http/Controllers/ProductController.php]: done

File: app/Http/Controllers/ProductController.php [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Validator;
+
+class ProductController extends Controller
+{
+    // Fetch all products
+    public function index(Request $request): JsonResponse
+    {
+        $validator = Validator::make($request->all(), [
+            'page' => 'integer|min:0',
```
Comment (BinyamMamo) [inline app/Http/Controllers/ProductController.php:16]: page should start from 1, right?

File: app/Http/Controllers/ProductController.php [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            return response()->json([
+                'message' => 'Validation failed',
+                'errors' => $validator->errors(),
+            ], 422);
+        }
+
+        $fields = $validator->validated();
+        $fields['per_page'] = $fields['per_page'] ?? 10;
+
+        return response()->json(Product::paginate($fields['per_page']));
```
Comment (BinyamMamo) [inline app/Http/Controllers/ProductController.php:30]: shouldn't this method take the `page` field as an argument

File: app/Http/Controllers/ProductController.php [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            ], 422);
+        }
+
+        $fields['vendor_id'] = $vendor->id;
+
+        $imageUrls = [];
+
+        if ($request->hasFile('images')) {
+            foreach ($request->file('images') as $image) {
+                $uploadedFileUrl = cloudinary()->upload($image->getRealPath())->getSecurePath();
```
Comment (BinyamMamo) [inline app/Http/Controllers/ProductController.php:76]: don't we need the public id for these too

File: app/Http/Controllers/ProductController.php (hunk 3 above)
Comment (carnage-mode) [reply app/Http/Controllers/ProductController.php:16]: Oh yeah, noted

File: app/Http/Controllers/ProductController.php (hunk 4 above)
Comment (carnage-mode) [reply app/Http/Controllers/ProductController.php:30]: I believe it takes in the items to show in a page. The page field is automatically taken from the request. You can test this by adding `?page=[number]`

File: app/Http/Controllers/ProductController.php (hunk 5 above)
Comment (carnage-mode) [reply app/Http/Controllers/ProductController.php:76]: Noted

Comment (carnage-mode) [reply app/Http/Controllers/ProductController.php:76]: One issue with us using an array of images here, we can't really add another field for the public id and guarantee that both exist and correspond to eachother. I wish there was something like a hashmap or a tuple in sql that we can use for this, but I'm not aware of anything

Comment (carnage-mode) [reply app/Http/Controllers/ProductController.php:76]: The best solution would be to create a Model for images and just put all of our images there. That'd also allow us to make generalized helpers for cloudinary stuff. For now let's just ignore this and rely on the URLs only.

Comment (carnage-mode) [reply app/Http/Controllers/ProductController.php:76]: Is that good with you @BinyamMamo

**Pull request merged**